statement_test.go

```diff
@@ -0,0 +1,296 @@
+	var db *sql.DB
+	var err error
+
+	if db, err = sql.Open("snowflake", dsn); err != nil {
+		t.Fatalf("failed to open db. %v, err: %v", dsn, err)
+	}
+	defer db.Close()
+
+	_, err = db.Exec("drop table if exists test_multi_statement_txn")
+	if err != nil {
+		t.Fatalf("failed to drop table: %v", err)
+	}
+
+	_, err = db.Exec("create or replace table test_multi_statement_txn(c1 number, c2 string)" +
+		"as select 10, 'z'")
+	if err != nil {
+		t.Fatalf("failed to create table: %v", err)
+	}
+
+	ctx, _ := WithMultiStatement(context.Background(), 4)
+	multiStmtQuery := "begin;\n" +
+		"delete from test_multi_statement_txn;\n" +
+		"insert into test_multi_statement_txn values (1, 'a'), (2, 'b');\n" +
+		"commit;"
+	_, err = db.QueryContext(ctx, multiStmtQuery)
+	if err != nil {
+		t.Fatalf("failed to query multiple statements: %v", err)
+	}
+
+	_, err = db.Exec("drop table if exists test_multi_statement_txn")
+	if err != nil {
+		t.Fatalf("failed to drop table: %v", err)
+	}
+}
+
+func TestMultiStatementExecuteMix(t *testing.T) {
+	var db *sql.DB
+	var err error
+
+	if db, err = sql.Open("snowflake", dsn); err != nil {
+		t.Fatalf("failed to open db. %v, err: %v", dsn, err)
+	}
+	defer db.Close()
+
+	ctx, _ := WithMultiStatement(context.Background(), 3)
+	multiStmtQuery := "create or replace temporary table test_multi (cola int);\n" +
+		"insert into test_multi values (1), (2);\n" +
+		"select cola from test_multi order by cola asc;"
+	res, err := db.ExecContext(ctx, multiStmtQuery)
+	if err != nil {
+		t.Fatal("failed to execute statement: ", err)
+	}
+
+	count, err := res.RowsAffected()
+	if err != nil {
+		t.Fatalf("res.RowsAffected() returned error: %s", err.Error())
+	}
+	if count != 2 {
+		t.Fatalf("expected 3 affected rows, got %d", count)
+	}
+}
+
+func TestMultiStatementQueryMix(t *testing.T) {
+	var db *sql.DB
+	var err error
+
+	if db, err = sql.Open("snowflake", dsn); err != nil {
+		t.Fatalf("failed to open db. %v, err: %v", dsn, err)
+	}
+	defer db.Close()
+
+	ctx, _ := WithMultiStatement(context.Background(), 3)
+	multiStmtQuery := "create or replace temporary table test_multi (cola int);\n" +
+		"insert into test_multi values (1), (2);\n" +
+		"select cola from test_multi order by cola asc;"
+	rows, err := db.QueryContext(ctx, multiStmtQuery)
+	if err != nil {
+		t.Fatal("failed to execute statement: ", err)
+	}
+	defer rows.Close()
+
+	// first statement
+	if !rows.Next() {
+		t.Error("failed to query")
+	}
+
+	var count, v int
+	// second statement
+	rows.NextResultSet()
+	if rows.Next() {
+		err = rows.Scan(&count)
+		if err != nil {
+			t.Errorf("failed to scan: %#v", err)
+		}
+		if count != 2 {
+			t.Fatalf("expected 2 affected rows, got %d", count)
+		}
+	}
+
+	expected := 1
+	// third statement
+	rows.NextResultSet()
+	for rows.Next() {
+		err = rows.Scan(&v)
+		if err != nil {
+			t.Errorf("failed to scan: %#v", err)
+		}
+		if v != expected {
+			t.Fatalf("failed to fetch. value: %v", v)
+		}
+		expected++
+	}
+}
```

Comment:
Overall positive test cases look ok.
Can you add negative tests? For example, setting 3 statements and run 2 or 4 queries and get the error?

Reply:
Add the test cases
- Set the user/account's MultiStatement parameter to 0 and run multiple queries.